### PR TITLE
Implement multiplayer support for minimap

### PR DESCRIPTION
**Pull request by darshan3**
**How to test:**
Connect using multiple players, you should be able to see each other's arrows on minimap when in range.

### CODE REVIEW

File: src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java [hunk 1]
```diff
@@ -16,9 +16,7 @@
 package org.terasology.minimap.rendering.nui.layers;
 
 
-import java.util.Collection;
-import java.util.Optional;
-import java.util.PriorityQueue;
+import java.util.*;
```
Comment (iaronaraujo) [inline src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java]: star import

File: src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java [hunk 2]
```diff
@@ -78,6 +87,9 @@ public EntityRef get() {
         int seaLevel = worldGenerator == null ? 10 : worldGenerator.getWorld().getSeaLevel();
         minimapHUDElement.setHeightRange(seaLevel, seaLevel + 64);
         minimapHUDElement.setWorldProvider(worldProvider);
+        if (entityManager.getCountOfEntitiesWith(AliveCharacterComponent.class) != 0) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java]: As `minimapHUDElement.updateAlivePlayerList` can easily deal with empty lists, I don't see why the guarding `if` expression is required here. 
In particular, I don't think that this initialize method will be called often (hopefully only once 🙄 ), so searching for all entities should be fine.

File: src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java [hunk 3]
```diff
@@ -125,4 +137,34 @@ public void onPlaceBlock(PlaceBlocks event, EntityRef entity) {
             minimapHUDElement.updateLocation(pos);
         }
     }
+
+    @ReceiveEvent //Spawn, initial spawn on joining a server
+    public void onPlayerSpawnEvent(OnPlayerSpawnedEvent event, EntityRef player) {
+        if (minimapHUDElement != null) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java]: Is this `null` check really necessary? I guess due it being a system we cannot make the member `final` and initialize it in the constructor... 
Can we live with the risk of it being `null`?

File: src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java [hunk 4]
```diff
@@ -125,4 +137,34 @@ public void onPlaceBlock(PlaceBlocks event, EntityRef entity) {
...
+    @ReceiveEvent //Spawn, initial spawn on joining a server
+    public void onPlayerReSpawnEvent(OnPlayerRespawnedEvent event, EntityRef player) {
+        if (minimapHUDElement != null) {
+            minimapHUDElement.addAlivePlayer(player);
+        }
+    }
+
+    @ReceiveEvent
+    public void onAwaitedLocalCharacterSpawnEvent(AwaitedLocalCharacterSpawnEvent event, EntityRef player) {
+        if (minimapHUDElement != null && entityManager.getCountOfEntitiesWith(AliveCharacterComponent.class) != 0) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java]: Again the question, do we actually need this check?

File: src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java [hunk 5]
```diff
@@ -84,6 +83,9 @@
     private Binding<EntityRef> targetEntityBinding = new DefaultBinding<>(EntityRef.NULL);
     private Binding<Integer> zoomFactorBinding = new DefaultBinding<>(0);
 
+    private Set<EntityRef> alivePlayers = new HashSet<EntityRef>();
+    private Logger logger = LoggerFactory.getLogger(MinimapGrid.class);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java]: Should be 
```suggestion
    static final Logger logger = LoggerFactory.getLogger(MinimapGrid.class);
```

File: src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java [hunk 6]
```diff
@@ -84,6 +83,9 @@
     private Binding<EntityRef> targetEntityBinding = new DefaultBinding<>(EntityRef.NULL);
     private Binding<Integer> zoomFactorBinding = new DefaultBinding<>(0);
 
+    private Set<EntityRef> alivePlayers = new HashSet<EntityRef>();
```
Comment (skaldarnar) [inline src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java]: Can be 
```suggestion
    private final Set<EntityRef> alivePlayers = new HashSet<EntityRef>();
```

Comment (darshan3) [reply src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java]: If we keep it final, can it be changed? More players will be added as the join the game.

File: src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java (hunk 5 above)
Comment (darshan3) [reply src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java]: Will remove logger altogether.

File: src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java (hunk 3 above)
Comment (darshan3) [reply src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java]: Sometimes the minimapHUDElement is not initialized and players are already getting spawned This mainly happens when a client joins the game, so the spawn event for other players already in the game is called.

File: src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java (hunk 1 above)
Comment (darshan3) [reply src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java]: Done.

File: src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java (hunk 6 above)
Comment (darshan3) [reply src/main/java/org/terasology/minimap/rendering/nui/layers/MinimapGrid.java]: I got confused between final and immutable. I will change it to final.

File: src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java (hunk 3 above)
Comment (skaldarnar) [reply src/main/java/org/terasology/logic/players/DefaultMinimapSystem.java]: what happens with the event in that case? will the respective player be added to the minimap in that case?

**Pull request merged**